File: 백준/Gold/13460. 구슬 탈출 2/구슬 탈출 2.py
from collections import deque


def solve(board):
    N = len(board)
    M = len(board[0])

    directions = [(-1, 0), (1, 0), (0, -1), (0, 1)]

    def move(x, y, dx, dy):
        count = 0
        while board[x + dx][y + dy] != '#' and board[x][y] != 'O':
            x += dx
            y += dy
            count += 1
        return x, y, count

    for i in range(N):
        for j in range(M):
            if board[i][j] == 'R':
                rx, ry = i, j
            if board[i][j] == 'B':
                bx, by = i, j

    queue = deque([(rx, ry, bx, by, 0)])
    visited = {rx, ry, bx, by}

    while queue:
        rx, ry, bx, by, depth = queue.popleft()

        if depth >= 10:
            return -1

        for dx, dy in directions:
            nrx, nry, red_moves = move(rx, ry, dx, dy)
            nbx, nby, blue_moves = move(bx, by, dx, dy)

            if board[nbx][nby] == 'O':
                continue
            if board[nrx][nry] == 'O':
                return depth + 1

            if (nrx, nry) == (nbx, nby):
                if red_moves > blue_moves:
                    nrx -= dx
                    nry -= dy
                else:
                    nbx -= dx
                    nby -= dy

            if (nrx, nry, nbx, nby) not in visited:
                visited.add((nrx, nry, nbx, nby))
                queue.append((nrx, nry, nbx, nby, depth + 1))

    return -1


N, M = map(int, input().rstrip().split())
board = [list(input()) for _ in range(N)]
print(solve(board))

File: 백준/Gold/13460. 구슬 탈출 2/구슬 탈출 2.cc
#include <bits/stdc++.h>
using namespace std;

int dx[4] = {-1, 1, 0, 0};
int dy[4] = {0, 0, -1, 1};

struct State
{
    int rx, ry;
    int bx, by;
    int moves;
};

int main()
{
    int N, M;
    cin >> N >> M;
    vector<string> board(N);
    for (auto &s : board)
        cin >> s;

    int rx, ry, bx, by, hole_x, hole_y;
    for (int i = 0; i < N; i++)
    {
        for (int j = 0; j < M; j++)
        {
            if (board[i][j] == 'R')
            {
                rx = i;
                ry = j;
                board[i][j] = '.';
            }
            if (board[i][j] == 'B')
            {
                bx = i;
                by = j;
                board[i][j] = '.';
            }
            if (board[i][j] == 'O')
            {
                hole_x = i;
                hole_y = j;
            }
        }
    }

    bool visited[10][10][10][10];
    memset(visited, false, sizeof(visited));
    queue<State> q;
    q.push(State{rx, ry, bx, by, 0});
    visited[rx][ry][bx][by] = true;
    int result = -1;

    while (!q.empty())
    {
        State current = q.front();
        q.pop();
        if (current.moves >= 10)
            continue;
        for (int dir = 0; dir < 4; dir++)
        {
            int nrx = current.rx, nry = current.ry;
            int nbx = current.bx, nby = current.by;
            bool red_in = false, blue_in = false;

            auto move_ball = [&](int &x, int &y) -> bool
            {
                while (true)
                {
                    if (board[x + dx[dir]][y + dy[dir]] == '#')
                        break;
                    x += dx[dir];
                    y += dy[dir];
                    if (x == hole_x && y == hole_y)
                        return true;
                }
                return false;
            };

            bool first_red = false;
            if (dir == 0)
            {
                if (current.rx < current.bx)
                    first_red = true;
            }
            else if (dir == 1)
            {
                if (current.rx > current.bx)
                    first_red = true;
            }
            else if (dir == 2)
            {
                if (current.ry < current.by)
                    first_red = true;
            }
            else if (dir == 3)
            {
                if (current.ry > current.by)
                    first_red = true;
            }

            if (first_red)
            {
                red_in = move_ball(nrx, nry);
                blue_in = move_ball(nbx, nby);
            }
            else
            {
                blue_in = move_ball(nbx, nby);
                red_in = move_ball(nrx, nry);
            }

            if (blue_in)
                continue;
            if (red_in)
            {
                result = current.moves + 1;
                cout << result;
                return 0;
            }

            if (nrx == nbx && nry == nby)
            {
                int red_dist = abs(nrx - current.rx) + abs(nry - current.ry);
                int blue_dist = abs(nbx - current.bx) + abs(nby - current.by);
                if (red_dist > blue_dist)
                {
                    nrx -= dx[dir];
                    nry -= dy[dir];
                }
                else
                {
                    nbx -= dx[dir];
                    nby -= dy[dir];
                }
            }

            if (!visited[nrx][nry][nbx][nby])
            {
                visited[nrx][nry][nbx][nby] = true;
                q.push(State{nrx, nry, nbx, nby, current.moves + 1});
            }
        }
    }

    cout << result;
}